File: seminar-07/C/java/C.java
import java.io.*;
import java.util.*;

public class C {
    static final long INF = (long) 1e18;

    static class Edge {
        int to;
        long w;
        Edge(int to, long w) {
            this.to = to;
            this.w = w;
        }
    }

    // Пара (dist, v) + сравнение по dist, потом по v
    static class Node implements Comparable<Node> {
        long dist;
        int v;
        Node(long dist, int v) {
            this.dist = dist;
            this.v = v;
        }
        @Override
        public int compareTo(Node o) {
            if (this.dist < o.dist) return -1;
            if (this.dist > o.dist) return 1;
            return Integer.compare(this.v, o.v);
        }
    }

    static class MinHeap {
        private final ArrayList<Node> a = new ArrayList<>();

        boolean isEmpty() {
            return a.isEmpty();
        }

        void push(Node x) {
            a.add(x);
            shiftUp(a.size() - 1);
        }

        Node pop() {
            Node top = a.get(0);
            Node last = a.remove(a.size() - 1);
            if (!a.isEmpty()) {
                a.set(0, last);
                shiftDown(0, a.size());
            }
            return top;
        }

        private void shiftUp(int i) {
            while (i > 0) {
                int p = (i - 1) / 2;
                if (a.get(p).compareTo(a.get(i)) <= 0) break;
                Node tmp = a.get(p);
                a.set(p, a.get(i));
                a.set(i, tmp);
                i = p;
            }
        }

        private void shiftDown(int i, int n) {
            while (true) {
                int l = 2 * i + 1;
                if (l >= n) break;
                int r = l + 1;
                int j = l;
                if (r < n && a.get(r).compareTo(a.get(l)) < 0) j = r;
                if (a.get(i).compareTo(a.get(j)) <= 0) break;
                Node tmp = a.get(i);
                a.set(i, a.get(j));
                a.set(j, tmp);
                i = j;
            }
        }
    }

    static long[] dijkstra(ArrayList<Edge>[] adj, int start) {
        int n = adj.length;
        long[] dist = new long[n];
        Arrays.fill(dist, INF);
        dist[start] = 0;

        MinHeap q = new MinHeap();
        q.push(new Node(0, start));

        while (!q.isEmpty()) {
            Node cur = q.pop();
            long curDist = cur.dist;
            int v = cur.v;

            if (curDist != dist[v]) continue;

            for (Edge e : adj[v]) {
                int to = e.to;
                long nd = curDist + e.w;
                if (nd < dist[to]) {
                    dist[to] = nd;
                    q.push(new Node(nd, to));
                }
            }
        }
        return dist;
    }

    public static void main(String[] args) throws Exception {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

        String[] first = in.readLine().split(" ");
        int n = Integer.parseInt(first[0]);
        int m = Integer.parseInt(first[1]);

        @SuppressWarnings("unchecked")
        ArrayList<Edge>[] adj = new ArrayList[n];
        for (int i = 0; i < n; i++) adj[i] = new ArrayList<>();

        for (int i = 0; i < m; i++) {
            String[] parts = in.readLine().split(" ");
            int fr = Integer.parseInt(parts[0]);
            int to = Integer.parseInt(parts[1]);
            long w = Long.parseLong(parts[2]);
            adj[fr].add(new Edge(to, w));
        }

        long[] dist = dijkstra(adj, 0);

        StringBuilder out = new StringBuilder();
        for (int i = 1; i < n; i++) {
            out.append(dist[i]).append('\n');
        }
        System.out.print(out.toString());
    }
}

File: seminar-07/C/python/C.py
INF = 10**18

class MinHeap:
    def __init__(self, data=None):
        self.a = []
        if data is not None:
            self.a = list(data)
            self.heapify()

    def __len__(self):
        return len(self.a)

    def is_empty(self):
        return len(self.a) == 0

    def peek(self):
        if not self.a:
            return None
        return self.a[0]

    def push(self, x):
        self.a.append(x)
        self.shift_up(len(self.a) - 1)

    def pop(self):
        if not self.a:
            return None
        top = self.a[0]
        last = self.a.pop()
        if self.a:
            self.a[0] = last
            self.shift_down(0)
        return top

    def shift_up(self, i):
        a = self.a
        while i > 0:
            p = (i - 1) // 2
            if a[p] <= a[i]:
                break
            a[p], a[i] = a[i], a[p]
            i = p

    def shift_down(self, i, n=None):
        a = self.a
        if n is None:
            n = len(a)
        while True:
            l = 2 * i + 1
            if l >= n:
                break
            r = l + 1
            j = l
            if r < n and a[r] < a[l]:
                j = r
            if a[i] <= a[j]:
                break
            a[i], a[j] = a[j], a[i]
            i = j

    def heapify(self):
        n = len(self.a)
        for i in range(n // 2 - 1, -1, -1):
            self.shift_down(i, n)


def dijkstra(adj_list, start):
    n = len(adj_list)
    dist = [INF] * n
    dist[start] = 0

    q = MinHeap()
    q.push((0, start))  # (расстояние, вершина)

    while not q.is_empty():
        cur_dist, cur = q.pop()

        # устаревшая запись (в кучу могли положить более плохое расстояние ранее)
        if cur_dist != dist[cur]:
            continue

        for nxt, w in adj_list[cur]:
            nd = cur_dist + w
            if nd < dist[nxt]:
                dist[nxt] = nd
                q.push((nd, nxt))

    return dist

n, m = map(int, input().split())
adj_list = [[] for _ in range(n)]
for _ in range(m):
    fr, to, w = map(int, input().split())
    adj_list[fr].append((to, w))
dist = dijkstra(adj_list, 0)
print(*dist[1:], sep="\n")